### ISSUE TITLE
use internal definition contexts instead of lambda wrapper

### ISSUE THREAD

**Issue by stchang** [stchang/macrotypes#31]
cc @iitalics @michaelballantyne

Opening this issue to hopefully avoid some duplication of effort, eg [1] and [2]

I also tried this a long time ago but could not get it to work, in part because Turnstile sometimes expands unbound type variables (eg, in `assign-type`), which the macro expander currently allows.

@michaelballantyne investigated further and found that the current implementation "works" because `#%top` is bound where the type var is referenced. If the type var is replaced with a `generate-temporary`, then an unbound id error is produced instead.

(Also, one of the tests was incorrect: https://github.com/michaelballantyne/macrotypes/commit/f9a8113fad952d8b995d378d65567f6805407ad0)

[1]: https://github.com/iitalics/macrotypes/tree/int-defn-ctx
[2]: https://github.com/michaelballantyne/macrotypes/tree/intdef

**Comment by iitalics** [comment]
Thanks. I was about to ask you about this, because it seems like the current bound vars work by magic (they are pulled from the expanded lambda, then re-expanded in a new context, yet they dont throw unbound id errors). You're saying that they have some #%top in scope somehow?

**Comment by iitalics** [comment]
Could we somehow inject a #%top into their scope to emulate their behavior? I would really like to use this, as it would free up the ability to use stop-ids in local-expand if we can get it working

**Comment by stchang** [comment]
Yes I guess we could, but now I'm not sure if we always want the current behavior. For example, [one of the tests in `exist.rkt`][1] was incorrectly *not* throwing unbound id error because of this. Did you run into problems with these tests?

[1]: https://github.com/michaelballantyne/macrotypes/commit/f9a8113fad952d8b995d378d65567f6805407ad0

**Comment by michaelballantyne** [comment]
My understanding was incorrect; it has nothing to do with #%top. local-expand never applies implicit #%top anyway.

The exists test that was wrong was a situation where the type variable had a binding, but the binding was out of context. With full expansion that produces an unbound identifier error:

```
(define-syntax (m stx)
  (define ref (syntax-case (local-expand #`(let ([Y 5]) Y)
                                         'expression
                                         '())
                ()
                [(_ _ b) #'b]))
  ref)
(m)
; Y: identifier used out of context
;  context...:
;  matching binding...: in: Y
```

But with local-expand, it does not:

```
(define-syntax (m stx)
  (define ref (syntax-case (local-expand #`(let ([Y 5]) Y)
                                         'expression '())
                ()
                [(_ _ b) #'b]))
  (local-expand ref 'expression '())
  #'(void))
(m)
```

runs without error. I believe this is intended to support cases where you disassemble syntax, local-expand it, and reassemble it.

The result is that even though the existential variable was not bound in the final expansion of the type, it expanded fine because it had had a binding previously and local-expand does not produce identifier out of context errors.

I ran across that test for another reason, which is that syntax-local-introduce wasn't used in all the right places. This didn't cause a problem when the bindings create by `infer` used use-site identifiers, but does when it creates bindings with macro introduced identifiers.

Right now turnstile calls syntax-local-introduce whenever is stashes something in a syntax property. But it is also necessary to do the same when retrieving syntax out of a property. See my fix here: https://github.com/michaelballantyne/macrotypes/blob/7f1e534e0fd69538f1da188f8a82dde034ce3764/macrotypes/typecheck.rkt#L213-L228

I have internal definition contexts in my branch working now for everything except the dependent type tests. Not sure what's going on there yet, but maybe something related to `subst`? I did have to avoid generating a new symbolic name for the identifiers bound by infer, because that seemed to break a bunch of things (something that looked like typeclasses, for one).

Was planning to dig into those two issues next week, but I don't want to duplicate work. @iitalics, do you want to push on this or should I? How well is your version working?

For context, Matthias has me working on fixing turnstile's quadratic expansion time problem. I plan to have typed forms expand to something in the stop list so they won't re-expand until we're all the way out of typechecking. So I need expansion with a stop list to work.

**Comment by iitalics** [comment]
My version isn't working very well, I just wrote it last night. It passes STLC but I stopped working on it while debugging sysf.

I'm going to look over your work; you're definitely much more knowledgable than me on local-expand etc.

>  I plan to have typed forms expand to something in the stop list so they won't re-expand until we're all the way out of typechecking.

I agree that we should do something like this. I was also crafting an example elaboration-target language that breaks when typed forms are fully-expanded.

EDIT: [here it is](https://github.com/stchang/macrotypes/compare/master...iitalics:sm), although it's not extremely relevant to using internal definition contexts
